cpp/src/arrow/array/builder_nested.cc

```diff
@@ -99,6 +99,11 @@ Status ListBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
     RETURN_NOT_OK(value_builder_->FinishInternal(&items));
   }
 
+  // If the type has not been specified in the constructor, infer it
+  if (!arrow::internal::checked_cast<ListType&>(*type_).value_type()) {
```

Comment:
Ah, this is for the case where type_.value_type() has been initialized to nullptr in the constructor (i.e. the type wasn't known when the ListBuilder was constructed so it was nullptr there), so it will be computed in the Finish method. This is used if the type is inferred automatically during Finish. This requires btw that .Finish of the value builder is called first (which it is). Does that make sense?

Reply:
Ahh.... Do you mean when `value_builder_` itself is a union type? If so, can you add a comment? :-) Thanks.

Reply:
added the comment :)